src/build/cargo.rs

```diff
@@ -299,20 +308,29 @@ impl Executor for RlsExecutor {
             }
         }
 
-        if self.workspace_mode {
-            let output = cmd.output().expect("Couldn't execute rustc");
-            let mut stderr_json_msg = convert_message_to_json_strings(output.stderr);
-            self.compiler_messages.lock().unwrap().append(&mut stderr_json_msg);
-        } else {
-            cmd.status().expect("Couldn't execute rustc");
-        }
-
         // Finally, store the modified cargo-generated args/envs for future rustc calls
         args.insert(0, rustc_exe);
         let mut compilation_cx = self.compilation_cx.lock().unwrap();
         compilation_cx.args = args;
         compilation_cx.envs = cargo_cmd.get_envs().clone();
 
+        if self.workspace_mode {
+            let args = &compilation_cx.args;
+            let envs = &compilation_cx.envs;
+            let build_dir = compilation_cx.build_dir.as_ref().unwrap();
+
+            match super::rustc::rustc(&self.vfs, args, envs, build_dir, self.config.clone()) {
+                BuildResult::Success(mut messages, mut analysis) |
+                BuildResult::Failure(mut messages, mut analysis) => {
+                    self.compiler_messages.lock().unwrap().append(&mut messages);
+                    self.analyses.lock().unwrap().append(&mut analysis);
+                }
+                _ => {}
+            }
+        } else {
+            cmd.status().expect("Couldn't execute rustc");
```

Comment:
I see that I made it so the lock on compilation_cx is unnecessarily held for the whole duration of compilation, I'll fix that.

Besides that, should we set it after in case `cmd.status()` or `rustc::rustc()` panics and we don't store bogus values from faulty compilation? If not, then I think this may not be a problem, since we feed the values directly to `rustc::rustc` and it doesn't touch it nor does `cmd.status()` require it.